Css fixes

deerman31 opened the pull request:
1. tozeki さんによる CSS の修正を適用
2. addNewTransaction の 4 つの入力フィールドにバリデーション処理を追加
3. 各リストの表示高さを調整
4. 入力されたSOLをスマートコントラクトに渡すためにlamportsの単位に変換するようにしました。

これらの変更により、UIの改善とデータ入力の信頼性向上を図りました。

deerman31 committed:
リストの高さを調整

deerman31 committed:
リストの表示を修正

deerman31 committed:
AddNewTransactionの入力のvalidateする関数を追加

deerman31 committed:
tozekiさんのcssの修正を追加

deerman31 committed:
solからlamportsに変換する処理を追加

deerman31 committed:
fmtをし、lintが通るようにした

rikeda-cloud commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,24 +11,88 @@ interface input {
+    return false;
+  }
+  refundDeadline = refundDeadline.trim();
+  if (transactionInfo.length > 0 && transactionInfo.length < 51) {
+    alert('Error: transaction info');
+    return false;
+  }
```
判定条件が逆？

deerman31 replied:
ほんまや、初歩的なミスでした。

rikeda-cloud commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,24 +11,88 @@ interface input {
+  // 空文字列の場合はfalse
+  if (str.length === 0) return false;
+  // 負の場合と0の場合にfalse
+  if (str[0] === '-' || (str[0] === '0' && str.length === 1)) return false;
+  // 数値に変換してみる
+  const num = Number(str);
+  return !isNaN(num) && isFinite(num);
```
"00"を許容

t-ozzy reviewed (dismissed):
主にCSSの修正箇所を確認しました
format修正していただき、ありがとうございます

rikeda-cloud reviewed (changes requested):
より細かなバグは自分がテストする際に見ようかなと思います。
2箇所バリデーションについてコメントしたので確認お願いします。

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,24 +11,88 @@ interface input {
+  return num >= 1 && num <= 360;
+}
+
+function validateAddNewTransaction(
```
この関数名だと、trueかfalseかどちらが正しいinputなのか確証が持てないです。
以下のような命名はどうでしょうか
```ts
isTransactionValid()
isTxInfoValid()
```

deerman31 replied:
validateAddNewTransaction()で一気にvalidationせずに各入力値を変換する際に一緒にvalidationするように変更しました。

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -39,7 +104,8 @@ const AddNewTransactionComponent: React.FC<input> = ({
       num_of_transactions + 1,
-      Number(amount),
+      //Number(amount) * 10 ** 9,
+      Number(solToLamports(amount)),
       Number(refundDeadline) * 24 * 60 * 60,
```
jsでNumber型は浮動小数点数みたいなので、値が非常に大きい場合、
精度の問題がありそうです。
少なくとも以下は、浮動小数点の計算になってしまっていると思います。
BigNumber型ではだめですか？
```ts
      Number(refundDeadline) * 24 * 60 * 60,
```

deerman31 replied:
amountの部分の型に関しては全てBigNumberに変更しました。

tachu2 commented on srcs/app/frontend/src/app/components/Home/Body/Buyer/UnrecoverableList.tsx:
```diff
@@ -103,6 +103,10 @@ const UnrecoverableList = () => {
   }, [publicKey]);
 
+  const formatAmount = (lamports: bigint): string => {
+    return (Number(lamports) / 1e9).toFixed(9);
+  };
```
こちらも浮動小数点になってしまっています。
特にお金の計算に使っている場合は、注意が必要です。

tachu2 reviewed (commented):
対応ありがとうございます！

deerman31 committed:
api.tsとAddNewTransactionの部分でnumberの部分をBigNumberにしたり、validationを変換時に一緒に行うようにした。

deerman31 committed:
リストを表示する部分で精度に不安があったため、lamportsをsolに変換する部分で型にBigNumberを使用。

deerman31 committed:
fmtをかけました

deerman31 committed:
solToLamports()で最大値と最小値を決めそれから外れた値が来た際は弾くように変更

deerman31 committed:
fmtとlintを実行

deerman31 committed:
コンフリクトがあったため、以下に変更を加えていたが、mainのファイルをコピペしました。
以下のファイルへの変更はマージ後にブランチを分けて、再度変更を加えます。

Buyer/ClaimsList.tsx
UnrecoverableList.tsx
ClaimedRightsList.tsx
ClaimsExpiredList.tsx
api.ts

deerman31 committed:
fmtとlintを実行

deerman31 committed:
sampleに修正予定のファイルを追加。
api.ts
AddNewTransactionComponent.tsx
ClaimsList.tsx
UnrecoverableList.tsx
ClaimedRightsList.tsx
ClaimsExpiredList.tsx

deerman31 committed:
lintとfmtを通しました。

deerman31 committed:
コンフリクトを修正

deerman31 committed:
sampleのファイルをコンパイルなどの対象から除外

deerman31 committed:
テストを通すためにsampleに入れたファイルを削除

deerman31 committed:
Merge branch 'main' into css-Fixes

deerman31 committed:
修正を施したファイルを追加

tachu2 reviewed (commented):
昨日reviewしたバリデーションの箇所の変更が反映されていないですか？

deerman31 committed:
動かなかったので微修正を施した微修正を施した

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,182 @@ interface input {
+      error: 'Please enter a number of days between 1 and 360',
+    };
+  }
+  // 小数点以下を切り捨て
+  const roundedDays = dayNumber.integerValue(BigNumber.ROUND_DOWN);
```
この時点で小数点以下を考慮する必要があるんですか？
これの上でinputの範囲を確認しているように見えます。

deerman31 replied:
冗長なようなので削除しておきます。

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,182 @@ interface input {
+      alert(`Error: ${trInfo.error}`);
+      return;
+    }
+    const num_of_transactions = await countTransactions(
```
num_of_transactionsだけsnakecaseになっています。

deerman31 replied:
修正します

rikeda-cloud replied:
多分、少し前に自分が記述した箇所だったと思います。
すみません。

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,182 @@ interface input {
+boolを返すのでも良かったが、下の関数群と合わせるためにResultを採用 */
+function validateTransactionInfo(transactionInfo: string): Result<string> {
+  transactionInfo = transactionInfo.trim();
+  if (transactionInfo.length === 0 || transactionInfo.length > 50) {
```
`transactionInfo.length > 50`
lengthの値が50になっていますが、100ではなかったでしたっけ？
定数が定義されていたと思うので、それを使ってマジックナンバーを避けた方がいいと思います。

deerman31 replied:
ここは大まかな値を置いただけですが、
その定数がどこで定義されているのかわからないので、自分で定義しちゃいますね

rikeda-cloud replied:
ちなみに、スマートコンストラクタ側での受け入れは文字数ではなく、byte数なので、
3bytesの文字('あ')を34文字入力すると、フロントエンドでは成功するが、バックエンドでは失敗するような挙動があります

deerman31 replied:
> ちなみに、スマートコンストラクタ側での受け入れは文字数ではなく、byte数なので、 3bytesの文字('あ')を34文字入力すると、フロントエンドでは成功するが、バックエンドでは失敗するような挙動があります

半角英数字のみに絞っちゃいますかね。

rikeda-cloud replied:
良いと思います

tachu2 reviewed (commented):
気になった点をコメントしました。

YungTatyu commented:
僕の環境の問題なのか分からないのですが、frontの見た目がおかしく見えます。
JinやPlease connect your wallet to access buyer features.のような文字が白色になっていて、文字が全然みえないです。

rikeda-cloud commented:
> 僕の環境の問題なのか分からないのですが、frontの見た目がおかしく見えます。 JinやPlease connect your wallet to access buyer features.のような文字が白色になっていて、文字が全然みえないです。

自分の環境では文字の色は黒色です。
校舎Macで作業していたときにkusanoさんも同じ現象が起きていたと思います

deerman31 committed:
レビュー時に指摘された点などを修正

rikeda-cloud commented on srcs/app/frontend/src/app/components/api.ts:
```diff
@@ -101,7 +100,7 @@ export async function settleTransaction(
       .transaction();
 
-    const { blockhash } = await CONNECTION.getLatestBlockhash();
+    const { blockhash } = await CONNECTION.getRecentBlockhash();
```
getLatestBlockhashが正しいです。
恐らく、mainブランチのコンフリクト解消時に戻してしまったと思われます。
```
const { blockhash } = await CONNECTION.getLatestBlockhash();
```

deerman31 replied:
すでに修正し、pushしているのでそちらの差分は間違いだと思います。
現在修正したブランチのvscodeの画像です。

<img width="1440" alt="スクリーンショット 2024-10-04 14 42 50" src="https://github.com/user-attachments/assets/f4abb3a6-7338-48da-9d41-f5103600d0d6">

const { blockhash } = await CONNECTION.getLatestBlockhash();

こちらでも挙動確認でreturnなどが動かなかったので修正しました。

deerman31 committed:
fmtとlintをする

deerman31 committed:
amountの入力欄に(SOL)と単位を明示。

rikeda-cloud reviewed (dismissed):
私が指摘した部分のすべての修正を確認しました。
修正ありがとうございました！
LGTM!

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,207 @@ interface input {
+function validateTransactionInfo(transactionInfo: string): Result<string> {
+  transactionInfo = transactionInfo.trim();
+
+  if (!/^[a-zA-Z0-9]+$/.test(transactionInfo)) {
```
これは半角の記号も許容していますか？
これらは許容してあげたいです。
例えば
`;`,`:`, `,`
理由としては、サービス提供者側で、データの区切りで使うか可能性が高いからです。
例
```
netflix, standard plan, price, customer name
```

deerman31 replied:
,だけ許容するように修正します。

tachu2 replied:
sellerが自由に定義できるフィールドなので、他の半角記号や空白も許容してあげたいです。
半角のptintableなら許容したいです。

deerman31 replied:
> sellerが自由に定義できるフィールドなので、他の半角記号や空白も許容してあげたいです。 半角のptintableなら許容したいです。

半角英数と記号を全て許容するようにしました。

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,207 @@ interface input {
+  ) {
+    return {
+      value: '',
+      error: 'Transaction info should be no longer than 100 characters',
```
error: 'Transaction info should be no longer than 100 characters',
ここの文字数に定数を使うといいと思います。
以下のような感じでできると思います。
```ts
error: `Transaction info should be no longer than ${MAX_TRANSACTION_INFO} characters`
```

deerman31 replied:
修正します。

deerman31 committed:
半角英数と記号を許容するように修正

tachu2 commented on srcs/app/frontend/src/app/components/AddNewTransaction/AddNewTransactionComponent.tsx:
```diff
@@ -10,38 +11,207 @@ interface input {
+function validateTransactionInfo(transactionInfo: string): Result<string> {
+  transactionInfo = transactionInfo.trim();
+
+  if (!/^[\x20-\x7E]+$/.test(transactionInfo)) {
```
これだけだと何を意味しているかわからないので、コメントがあるべきだとおもいます。
